compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -154,13 +154,13 @@ pub struct UnifyReceiverContext<'tcx> {
 
 #[derive(Clone, PartialEq, Eq, Default, HashStable)]
 #[derive(TypeVisitable, TypeFoldable, TyEncodable, TyDecodable)]
-pub struct InternedObligationCauseCode<'tcx> {
+pub struct MaybeObligationCauseCode<'tcx> {
     /// `None` for `ObligationCauseCode::Misc` (a common case, occurs ~60% of
     /// the time). `Some` otherwise.
     code: Option<Arc<ObligationCauseCode<'tcx>>>,
 }
```

Comment:
I feel like `MaybeObligationCauseCode` is very misleading. This is always an `ObligationCauseCode` as `None` just represents `Arc::new(ObligationCauseCode::Misc)`.

`InternedObligationCauseCode` is just a more efficient way to store `ObligationCauseCode`. MAybe something like `ObligationCauseCodeRef` or whatever? I also think using `Interned` in the name even if it doesn't use proper *interning* is fine if there's no better name

Reply:
`ObligationCauseCodeHandle`?  I don't like `Ref` because it's not a reference.

Reply:
`Handle` sgtm